### ISSUE TITLE
[feature request] Disable GL_SHADER_DISK_CACHE for older games/runners

### ISSUE THREAD

**Issue by telanus** [lutris/lutris#3920]
Commit https://github.com/lutris/lutris/commit/f575dbb38f9a2ad4fcca3f668cdf2839eab74f4b is great for modern games, but clog up your rom folder if you play older games, IE pre-PSX era games, as these games don't need a GLShader.

Shouldn't we add an option to disable GLShader per runner/game.

@strycore @tannisroot @danieljohnson2

**Comment by tannisroot** [comment]
I don't understand, how does it clog up the rom folder? All games on Linux in some way use shaders and they all need to store shader cache somewhere. The option just redirects the location for storing it to a dedicated, per-game folder. Those folders will take up as much space as they would have if they were all stored in a single location.

**Comment by telanus** [comment]
> All games on Linux in some way use shaders and they all need to store shader cache somewhere.

That's not really that true. For most "modern" games shaders is a thing. But for "older" games, those pre-y2k shaders, if they had them, was totally different than today's shaders.

Also, in my case, emulators tend to either use their own location, IE yuzu, modern system emulators, but others, like MAME, RetroArch, use shaders the old way, just a post-processing filter that's used to display the emulated game.

Something like air/sea battle on the creativison being emulated on MAME doesn't need glshaders.

**Comment by tannisroot** [comment]
Yes, older games didn't have shader and used fixed function, but nowadays, modern graphics drivers don't support fixed function in hardware, they convert fixed function to modern shader pipeline, so in the end modern and old games both utilize shaders.
>Also, in my case, emulators tend to either use their own location, IE yuzu, modern system emulators, but others, like MAME, RetroArch, use shaders the old way, just a post-processing filter that's used to display the emulated game.
No, they can't use their own location unless they specify this exact environment variable, which I doubt they do. This shader cache is generated by the Nvidia driver itself and has nothing to do with whatever caches other runners have.
> Something like air/sea battle on the creativison being emulated on MAME doesn't need glshaders.
Don't look at the name of the environment variable, in Nvidia's terminology, GL is anything graphics api, ranging from first OpenGLs to Vulkan.

**Comment by danieljohnson2** [comment]
I still have the code for the settings option for this, but @strycore did not seem to like the UI clutter. Or maybe it was my Croatian translation that offended/horrified him.

But if you want to do this automatically, you'd have to detect 'old' games somehow. How would you do that?

**Comment by telanus** [comment]
> No, they can't use their own location unless they specify this exact environment variable, which I doubt they do

Here is a list of modern emulators and their shader cache location: https://emulation.gametechwiki.com/index.php/Shader_caches 

And this is RetroArch's idea: https://www.retroarch.com/?page=shaders



> I still have the code for the settings option for this,

If this is added to the runners config it won't take up too much space. 

> This shader cache is generated by the Nvidia driver itself and has nothing to do with whatever caches other runners have.

On my system I have only three games that Nvidia creased a glcache in .nv. all the other games don't create any entry. Until Lutris added this commit.

**Comment by strycore** [comment]
>  Or maybe it was my Croatian translation that offended/horrified him.

I don't speak Croatian so it was certainly the added option!

We can have a sane way to configure/handle this while avoiding additional options in the client.

We don't have to create a cache folder per game, we can say some runners have a global shader cache in ~/.cache/lutris/shaders so they don't clutter game directories. Most runners outside of Wine and Linux may not use directories for storing games so we could have the default location be in the ~/.cache/lutris folder and override it for the linux / wine runners.

**Comment by danieljohnson2** [comment]
It sounds like the idea is that Linux and WINE runners count as 'modern', and all others are 'old'. ('Retro'? 'Vintage'?)

So maybe only these runners should mess with the GL shader environment variables. We don't need to redirect the others at all, do we?

No Croatian would be needed for this.

**Comment by strycore** [comment]
This has nothing to do with retro or vintage systems, it's only differentiating emulators vs native.
Wine is not an emulator, it allows to run Windows executable as native Linux application so they should be handled as such.
For emulators, it's the emulator itself that is the native executable so it doesn't make sense to keep one cache per emulated game, just keep one cache per emulator. Some emulators have their own shader cache but that's entirely different, it has nothing to do with the Nvidia driver.

I'm still not sure if these environment variables make a noticeable difference in performance so maybe we can just skip the whole thing for emulators. But if we decide to have a specific shader cache for everything then storing 1 cache per runner except wine and linux makes the most sense.

**Comment by danieljohnson2** [comment]
I suppose that makes some kind of sense; DOSBOX, if it has shaders at all, has only DOSBOX shaders- the individual games don't expose any shaders of their own, and they might all shares their runners shaders.

I'll look into it.

**Comment by strycore** [comment]
DOSBox doesn't really handle any 3D stuff even if it uses OpenGL to draw its window. But Dolphin and RPCS3 run 3D games that require a lot of shader compilation. In addition to the emulator's own shader cache, there might be some benefits to let the nvidia driver have its own dedicated cache for this emulator.

**Comment by danieljohnson2** [comment]
How about this? A per runner shader cache in .cache/lutris/shaders/<runner>, except for WINE and Linux which continue to use the game's own directory.

We could put the shader cache for WINE and Linux games in .cache/lutris/<runner>/<game>, too, I 'spose.

**Comment by strycore** [comment]
yeah, that sounds great!

I feel that we should keep the cache in the game folder for windows and linux games, I'm concerned about the disk space the cache might occupy on a system drive if we keep everything in the same location.

**Comment by tannisroot** [comment]
> > No, they can't use their own location unless they specify this exact environment variable, which I doubt they do
> 
> Here is a list of modern emulators and their shader cache location: https://emulation.gametechwiki.com/index.php/Shader_caches 
> 
> And this is RetroArch's idea: https://www.retroarch.com/?page=shaders
> 
> 
> 
> > I still have the code for the settings option for this,
> 
> If this is added to the runners config it won't take up too much space. 
> 
> > This shader cache is generated by the Nvidia driver itself and has nothing to do with whatever caches other runners have.
> 
> On my system I have only three games that Nvidia creased a glcache in .nv. all the other games don't create any entry. Until Lutris added this commit.

That shader cache is a separate and unrelated thing. GL_CACHE is created by the Nvidia driver for all graphical applications, be it Vulkan ot OpenGL for its own purposes.
Can I ask why exactly having it separate is an issue? Does it take too much ask? It really shouldn't if the game is simple

**Comment by tannisroot** [comment]
> DOSBox doesn't really handle any 3D stuff even if it uses OpenGL to draw its window. But Dolphin and RPCS3 run 3D games that require a lot of shader compilation. In addition to the emulator's own shader cache, there might be some benefits to let the nvidia driver have its own dedicated cache for this emulator.

No, there shouldn't be any, any Dolphin games will have their own unique converted shaders, and I suspect some games might encounter the shader cache size issue as well, which is the whole point of the option.

**Comment by tannisroot** [comment]
> yeah, that sounds great!
> 
> I feel that we should keep the cache in the game folder for windows and linux games, I'm concerned about the disk space the cache might occupy on a system drive if we keep everything in the same location.

It shouldn't make any difference. The only difference is the location it stores the files in. And making it per-runner eliminates the whole point of the option, because some games will clutter the shader cache location with its own cache and cause the nvidia driver to prune the cache, causing shader stutters.

**Comment by telanus** [comment]
> No, there shouldn't be any, any Dolphin games will have their own unique converted shaders, and I suspect some games might encounter the shader cache size issue as well, which is the whole point of the option.

Dolphin's shaders is in ~/.cache/dolphin-emu/Shaders/ so creating a shader dir in the game's directory is useless (in my view) as Dolphin will never look there, and as @strycore said "For emulators, it's the emulator itself that is the native executable so it doesn't make sense to keep one cache per emulated game". 

> Can I ask why exactly having it separate is an issue? Does it take too much ask? It really shouldn't if the game is simple

I've tested it with MAME, PPSSPP, RetroArch and FS-UAE, and for all these emulators Lutris's implementation create the same file for every game I tried. So for the (amiga)[https://www.mobygames.com/browse/games/amiga/] it would create 3790+ useless files, if you run every game, and for mame it's even worse, as it emulates a whole lot more systems, so  guesstimate of 20000+ if you tests mame thoroughly, and I'm low-balling, according to http://adb.arcadeitalia.net/statistiche.php we're talking "The titles stored in the database are 19.4 millions, of which 7.8 millions for machines and 11.6 millions for software." that's a lot of wasted space if we create the same file for every game.    


> I'm concerned about the disk space the cache might occupy on a system drive if we keep everything in the same location.

Won't be much. MAME's cache is 33.4kb, while FS-UAE is 3.3KB, PPSSPP is about 178kb and RetroArch is about 7.3kb.  and is only needed for the emulator and not every emulated game. Dolphin's cache is about 300kb.

So In conclusion this is the best option:

> A per runner shader cache in .cache/lutris/shaders/[runner], except for WINE and Linux which continue to use the game's own directory.

The only problem I foresee (and have run into) is when all your roms live in the same folder, instead of each rom living in it's own dir, if the shader cache changes IE Citra you get a lot of stutter when you change from one game to the next, as the cache needs to be recompiled each time. This can be negated, if for Citra (and maybe Dolphin, but I don't have a lot of Dolphin based games) you have .cache/lutris/shaders/citra/[gamename]

**Comment by danieljohnson2** [comment]
Hmmm. Do not Citra games have game directories?

Maybe Citra and Dolphin should be like WINE and Linux. Citra is awkward since it is a JSON runner, but I'm sure it can be done. But it easier if it does not need to have its own custom policy, different from the others.

**Comment by telanus** [comment]
> Hmmm. Do not Citra games have game directories?
If you use *.3ds they can live all together, *.cia files tend to be installed on a directory in the nand folder.

3ds tend to be cartridges, while cia files are "downloaded games"

**Comment by danieljohnson2** [comment]
Ahh, I think I get it. I haven't been able to run Citra because it Lutris does not provide those files.

**Comment by strycore** [comment]
Games run with Citra are NOT native. They do not know what a X86_64 CPU is, they don't know about Nvidia, they don't know about OpenGL (maybe they know GLES) and they certainly don't know about Vulkan.

Individual folder games are useful for NATIVE PC GAMES. It doesn't matter if the game is for Linux or Windows. They are still PC native. 3DS titles are NOT PC GAMES. I can't believe this is so hard to explain. In our Discord server, the confusion is even greater than it is here.

**Comment by danieljohnson2** [comment]
I think that the source of confusion is why native-ness -however defined- is the critical factor, rather than something like 'has per game shaders' or 'stutters due to shade cache thrashing'.

Of course, we'll stick with WINE+Linux vs everything else if that's your decision. That's no worse than the current shipping behavior, surely.

**Comment by strycore** [comment]
I don't really want this implemented as a per-runner cache because I said so but because it makes sense to everyone involved and is the logical thing to do.
@telanus already pointed out that the various caches generated for a given emulator were the exact same. That's because when you run a game for MAME or DOSBox, you don't make a cache for Street Fighter II or Dune, you're making one for MAME and DOSBox.
Any game that wasn't made to explicitly talk to a PC GPU doesn't need a cache made for PC GPUs, it just doesn't make any sense.
